aio_pika/connection.py

```diff
@@ -205,8 +205,9 @@ def __del__(self):
         return self
 
     async def __aexit__(self, exc_type, exc_val, exc_tb):
-        for channel in tuple(self._channels.values()):
-            await channel.close()
+        for channel in tuple(self._channels.values()):  # can change size
+            if not channel.is_closed:
+                await channel.close()
```

Comment:
Hmm... that looks pretty simple but when you have to do this is the user closes channel manually. So, it's reasonable, but I imagine this simpler than the current patch.

```python
while (self._channels):
    channel = self._channels.pop()
    if channel.is_closed:
        continue
    await channel.close()
```

It's might be a guarantee of avoiding another edge case, a user opening a new channel while `` await channel.close()`` performing.

Another way is adding a lock or condition and trigger it while close is starting. But the first will be better I guess.

Reply:
I just noticed that we have a potential memory leak as it was in https://github.com/mosquito/aio-pika/pull/285. What if the user opens the channels and closes them himself? Do we clear `_channels` from closed channels?

Reply:
Seems to be yes.

Reply:
Then why I got `Channel already closed` if we delete it from `_channels`?

Reply:
If add `print(sys.getrefcount(channel))` onto 211 line into connection then will be print `6`? Why so many references?

Reply:
If disable `gc` then objects don't delete. It's a memory leak.